File: mobile/test/calculator_anjuran_resolver_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:peka_stunting/features/deteksi_dini/domain/calculator_anjuran_resolver.dart';
import 'package:peka_stunting/features/deteksi_dini/models/calculator_anjuran_rule.dart';

void main() {
  const resolver = CalculatorAnjuranResolver();

  final rules = [
    CalculatorAnjuranRule(
      sortOrder: 1,
      metric: CalculatorAnjuranRule.metricBmi,
      threshold: 30,
      operator: CalculatorAnjuranRule.operatorGt,
      isDefault: false,
      label: 'Obesitas',
      slug: 'obese',
      anjuran: 'Anjuran obesitas',
    ),
    CalculatorAnjuranRule(
      sortOrder: 2,
      metric: CalculatorAnjuranRule.metricBmi,
      threshold: 25,
      operator: CalculatorAnjuranRule.operatorGt,
      isDefault: false,
      label: 'Gemuk',
      slug: 'overweight',
      anjuran: 'Anjuran gemuk',
    ),
    CalculatorAnjuranRule(
      sortOrder: 3,
      metric: CalculatorAnjuranRule.metricBmi,
      threshold: 18.5,
      operator: CalculatorAnjuranRule.operatorGt,
      isDefault: false,
      label: 'Normal',
      slug: 'normal',
      anjuran: 'Anjuran normal',
    ),
    CalculatorAnjuranRule(
      sortOrder: 4,
      metric: CalculatorAnjuranRule.metricBmi,
      threshold: null,
      operator: CalculatorAnjuranRule.operatorGt,
      isDefault: true,
      label: 'Kurus',
      slug: 'underweight',
      anjuran: 'Anjuran kurus',
    ),
  ];

  test('resolves overweight for bmi 26.5', () {
    final resolved = resolver.resolve(
      rules: rules,
      metric: CalculatorAnjuranRule.metricBmi,
      value: 26.5,
    );

    expect(resolved?.slug, 'overweight');
    expect(resolved?.label, 'Gemuk');
    expect(resolved?.anjuran, 'Anjuran gemuk');
  });

  test('resolves default underweight for bmi 17', () {
    final resolved = resolver.resolve(
      rules: rules,
      metric: CalculatorAnjuranRule.metricBmi,
      value: 17,
    );

    expect(resolved?.slug, 'underweight');
    expect(resolved?.label, 'Kurus');
  });

  test('resolves normal lila at threshold', () {
    final lilaRules = [
      CalculatorAnjuranRule(
        sortOrder: 1,
        metric: CalculatorAnjuranRule.metricLilaCm,
        threshold: 23.5,
        operator: CalculatorAnjuranRule.operatorGte,
        isDefault: false,
        label: 'Normal',
        slug: 'normal',
        anjuran: 'Anjuran normal',
      ),
      CalculatorAnjuranRule(
        sortOrder: 2,
        metric: CalculatorAnjuranRule.metricLilaCm,
        threshold: null,
        operator: CalculatorAnjuranRule.operatorLt,
        isDefault: true,
        label: 'Berisiko KEK',
        slug: 'at_risk',
        anjuran: 'Anjuran KEK',
      ),
    ];

    final resolved = resolver.resolve(
      rules: lilaRules,
      metric: CalculatorAnjuranRule.metricLilaCm,
      value: 23.5,
    );

    expect(resolved?.slug, 'normal');
    expect(resolved?.label, 'Normal');
  });

  test('resolves at risk lila below threshold', () {
    final lilaRules = [
      CalculatorAnjuranRule(
        sortOrder: 1,
        metric: CalculatorAnjuranRule.metricLilaCm,
        threshold: 23.5,
        operator: CalculatorAnjuranRule.operatorGte,
        isDefault: false,
        label: 'Normal',
        slug: 'normal',
        anjuran: 'Anjuran normal',
      ),
      CalculatorAnjuranRule(
        sortOrder: 2,
        metric: CalculatorAnjuranRule.metricLilaCm,
        threshold: null,
        operator: CalculatorAnjuranRule.operatorLt,
        isDefault: true,
        label: 'Berisiko KEK',
        slug: 'at_risk',
        anjuran: 'Anjuran KEK',
      ),
    ];

    final resolved = resolver.resolve(
      rules: lilaRules,
      metric: CalculatorAnjuranRule.metricLilaCm,
      value: 22.0,
    );

    expect(resolved?.slug, 'at_risk');
    expect(resolved?.anjuran, 'Anjuran KEK');
  });

  test('resolves anemia medium risk for 5 yes answers', () {
    final anemiaRules = [
      CalculatorAnjuranRule(
        sortOrder: 1,
        metric: CalculatorAnjuranRule.metricYesCount,
        threshold: 7,
        operator: CalculatorAnjuranRule.operatorGt,
        isDefault: false,
        label: 'Resiko Tinggi Anemia',
        slug: 'high_risk',
        anjuran: 'Anjuran tinggi',
      ),
      CalculatorAnjuranRule(
        sortOrder: 2,
        metric: CalculatorAnjuranRule.metricYesCount,
        threshold: 4,
        operator: CalculatorAnjuranRule.operatorGte,
        isDefault: false,
        label: 'Risiko Sedang Anemia',
        slug: 'medium_risk',
        anjuran: 'Anjuran sedang',
      ),
      CalculatorAnjuranRule(
        sortOrder: 3,
        metric: CalculatorAnjuranRule.metricYesCount,
        threshold: 1,
        operator: CalculatorAnjuranRule.operatorGte,
        isDefault: false,
        label: 'Risiko Rendah Anemia',
        slug: 'low_risk',
        anjuran: 'Anjuran rendah',
      ),
      CalculatorAnjuranRule(
        sortOrder: 4,
        metric: CalculatorAnjuranRule.metricYesCount,
        threshold: null,
        operator: CalculatorAnjuranRule.operatorGte,
        isDefault: true,
        label: 'Tidak ada resiko Anemia',
        slug: 'normal',
        anjuran: 'Anjuran normal',
      ),
    ];

    final resolved = resolver.resolve(
      rules: anemiaRules,
      metric: CalculatorAnjuranRule.metricYesCount,
      value: 5,
    );

    expect(resolved?.slug, 'medium_risk');
    expect(resolved?.anjuran, 'Anjuran sedang');
  });

  test('resolves height for age stunted z score', () {
    final rules = [
      CalculatorAnjuranRule(
        sortOrder: 1,
        metric: CalculatorAnjuranRule.metricZScore,
        indicator: CalculatorAnjuranRule.indicatorHeightForAge,
        threshold: -2,
        operator: CalculatorAnjuranRule.operatorLt,
        isDefault: false,
        label: 'Pendek (stunting)',
        slug: 'stunted',
        anjuran: 'Anjuran stunting',
      ),
      CalculatorAnjuranRule(
        sortOrder: 2,
        metric: CalculatorAnjuranRule.metricZScore,
        indicator: CalculatorAnjuranRule.indicatorHeightForAge,
        threshold: null,
        operator: CalculatorAnjuranRule.operatorGte,
        isDefault: true,
        label: 'Normal',
        slug: 'normal',
        anjuran: 'Anjuran normal',
      ),
    ];

    final resolved = resolver.resolve(
      rules: rules,
      metric: CalculatorAnjuranRule.metricZScore,
      value: -2.5,
      indicator: CalculatorAnjuranRule.indicatorHeightForAge,
    );

    expect(resolved?.slug, 'stunted');
  });
}

File: backend/tests/Unit/CalculatorAnjuranResolverTest.php
<?php

namespace Tests\Unit;

use App\Models\CalculatorAnjuranRule;
use App\Models\EducationContent;
use App\Models\EducationItem;
use App\Models\EducationMenu;
use App\Services\Screening\CalculatorAnjuranResolver;
use App\Support\CalculatorAnjuranDefaults;
use Database\Seeders\EducationTaxonomySeeder;
use Database\Seeders\RoleSeeder;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;

class CalculatorAnjuranResolverTest extends TestCase
{
	use RefreshDatabase;

	private CalculatorAnjuranResolver $resolver;

	protected function setUp(): void
	{
		parent::setUp();

		$this->seed(RoleSeeder::class);
		$this->seed(EducationTaxonomySeeder::class);
		$this->resolver = app(CalculatorAnjuranResolver::class);
	}

	public function test_resolves_overweight_for_bmi_26_5(): void
	{
		$content = $this->createBmiContentWithRules();

		$resolved = $this->resolver->resolve(
			$content,
			CalculatorAnjuranRule::METRIC_BMI,
			26.5,
		);

		$this->assertSame('overweight', $resolved->slug);
		$this->assertSame('Gemuk', $resolved->label);
		$this->assertStringContainsString('gemuk', strtolower($resolved->anjuran));
	}

	public function test_resolves_default_underweight_for_low_bmi(): void
	{
		$content = $this->createBmiContentWithRules();

		$resolved = $this->resolver->resolve(
			$content,
			CalculatorAnjuranRule::METRIC_BMI,
			17.0,
		);

		$this->assertSame('underweight', $resolved->slug);
		$this->assertSame('Kurus', $resolved->label);
	}

	public function test_resolves_normal_lila_at_threshold_for_age_over_17(): void
	{
		$content = $this->createLilaContentWithRules();

		$resolved = $this->resolver->resolve(
			$content,
			CalculatorAnjuranRule::METRIC_LILA_CM,
			23.5,
			CalculatorAnjuranRule::INDICATOR_AGE_GT_17,
		);

		$this->assertSame('normal', $resolved->slug);
		$this->assertSame('Selamat, status gizi relatif normal', $resolved->label);
	}

	public function test_resolves_at_risk_lila_below_threshold_for_age_15_to_17(): void
	{
		$content = $this->createLilaContentWithRules();

		$resolved = $this->resolver->resolve(
			$content,
			CalculatorAnjuranRule::METRIC_LILA_CM,
			21.0,
			CalculatorAnjuranRule::INDICATOR_AGE_15_17,
		);

		$this->assertSame('at_risk', $resolved->slug);
		$this->assertSame('Anda berisiko kekurangan gizi (KEK)', $resolved->label);
	}

	public function test_resolves_normal_lila_for_age_10_to_14_at_threshold(): void
	{
		$content = $this->createLilaContentWithRules();

		$resolved = $this->resolver->resolve(
			$content,
			CalculatorAnjuranRule::METRIC_LILA_CM,
			18.5,
			CalculatorAnjuranRule::INDICATOR_AGE_10_14,
		);

		$this->assertSame('normal', $resolved->slug);
	}

	public function test_resolves_stunted_height_for_age_z_score(): void
	{
		$content = $this->createNutritionalStatusContentWithRules();

		$resolved = $this->resolver->resolve(
			$content,
			CalculatorAnjuranRule::METRIC_Z_SCORE,
			-2.5,
			CalculatorAnjuranRule::INDICATOR_HEIGHT_FOR_AGE,
		);

		$this->assertSame('stunted', $resolved->slug);
		$this->assertSame('Pendek (stunting)', $resolved->label);
	}

	public function test_resolves_anemia_tiers_by_yes_count(): void
	{
		$content = $this->createAnemiaContentWithRules();

		$this->assertSame(
			'normal',
			$this->resolver->resolve($content, CalculatorAnjuranRule::METRIC_YES_COUNT, 0)->slug,
		);
		$this->assertSame(
			'low_risk',
			$this->resolver->resolve($content, CalculatorAnjuranRule::METRIC_YES_COUNT, 3)->slug,
		);
		$this->assertSame(
			'medium_risk',
			$this->resolver->resolve($content, CalculatorAnjuranRule::METRIC_YES_COUNT, 5)->slug,
		);
		$this->assertSame(
			'high_risk',
			$this->resolver->resolve($content, CalculatorAnjuranRule::METRIC_YES_COUNT, 8)->slug,
		);
	}

	private function createBmiContentWithRules(): EducationContent
	{
		$menu = EducationMenu::query()->where('slug', 'remaja-putri')->firstOrFail();
		$item = EducationItem::query()
			->where('menu_id', $menu->id)
			->where('slug', 'cek-imt')
			->firstOrFail();

		$content = $item->content;
		$content->anjuranRules()->delete();

		foreach (CalculatorAnjuranDefaults::bmiRules() as $rule) {
			$content->anjuranRules()->create($rule);
		}

		return $content->fresh(['anjuranRules']);
	}

	private function createLilaContentWithRules(): EducationContent
	{
		$menu = EducationMenu::query()->where('slug', 'remaja-putri')->firstOrFail();
		$item = EducationItem::query()
			->where('menu_id', $menu->id)
			->where('slug', 'cek-lila')
			->firstOrFail();

		$content = $item->content;
		$content->anjuranRules()->delete();

		foreach (CalculatorAnjuranDefaults::lilaRulesRemajaPutri() as $rule) {
			$content->anjuranRules()->create($rule);
		}

		return $content->fresh(['anjuranRules']);
	}

	private function createAnemiaContentWithRules(): EducationContent
	{
		$menu = EducationMenu::query()->where('slug', 'remaja-putri')->firstOrFail();
		$item = EducationItem::query()
			->where('menu_id', $menu->id)
			->where('slug', 'cek-risiko-anemia')
			->firstOrFail();

		$content = $item->content;
		$content->anjuranRules()->delete();

		foreach (CalculatorAnjuranDefaults::anemiaRules() as $rule) {
			$content->anjuranRules()->create($rule);
		}

		return $content->fresh(['anjuranRules']);
	}

	private function createNutritionalStatusContentWithRules(): EducationContent
	{
		$menu = EducationMenu::query()->where('slug', 'bayi-dan-balita')->firstOrFail();
		$item = EducationItem::query()
			->where('menu_id', $menu->id)
			->where('slug', 'periksa-status-gizi')
			->firstOrFail();

		$content = $item->content;
		$content->anjuranRules()->delete();

		foreach (CalculatorAnjuranDefaults::nutritionalStatusRules() as $rule) {
			$content->anjuranRules()->create($rule);
		}

		return $content->fresh(['anjuranRules']);
	}
}
